Support rolling update to 0 desired replicas

ironcladlou opened the pull request:
Support a desired replica count of 0 for the new RC. Users sometimes
want to roll out a new "inactive" template with the intent of scaling
it up manually later.

Related to https://github.com/openshift/origin/issues/5787

cc @nikhiljindal @kargakis

kargakis commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -208,9 +208,11 @@ func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
 		return err
 	}
-	// Further validation.
-	if maxUnavailable == 0 && maxSurge == 0 {
```
Shouldn't we continue requiring this check when desired isn't zero?

ironcladlou replied:
Yeah, you're right. I'm going to have to force update to fix my committer email address.

kargakis commented:
Fixes https://github.com/kubernetes/kubernetes/issues/17000

ironcladlou committed:
Support rolling update to 0 desired replicas

Support a desired replica count of 0 for the new RC. Users sometimes
want to roll out a new "inactive" template with the intent of scaling
it up manually later.

kargakis commented:
LGTM

ironcladlou commented:
@nikhiljindal please review?

deads2k commented:
@k8s-bot test this please

deads2k commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -208,9 +208,15 @@ func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
+	// the entire scale of the old RC regardless of the configuration
+	// (equivalent to 100% maxUnavailable).
+	if desired == 0 {
+		maxUnavailable = original
```
I would have expect this to be equal to `max(spec.replicas, status.replicas)`.  I thought the `originalReplicasAnnotation` tracked previous state.

ironcladlou replied:
I think you're right.

ironcladlou replied:
Fixed

deads2k commented:
@k8s-bot unit test this please

deads2k commented:
lgtm

@ironcladlou you want to wait for @nikhiljindal  before merge?

ironcladlou commented:
Yeah, I'd like @nikhiljindal to vet the concept.

nikhiljindal commented:
@ironcladlou Can you please post a command that was failing before this change and is fixed by this PR?

I tried `kubectl create -f deployment.yaml` with replicas = 0 in that deployment and it worked fine without this change.

ironcladlou commented:
@nikhiljindal Try it with 0 replicas and a percentage based maxSurge/maxUnavailable (so that maxSurge and maxUnavailable are ultimately calculated as 0 relative to the replica count). I think that using an absolute value for maxSurge might work around the issue.

nikhiljindal commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -208,9 +208,15 @@ func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
+	// the effective scale of the old RC regardless of the configuration
+	// (equivalent to 100% maxUnavailable).
+	if desired == 0 {
+		maxUnavailable = int(math.Max(float64(oldRc.Spec.Replicas), float64(oldRc.Status.Replicas)))
```
What we really want is minAvailable to be zero on line 233, so I think setting this to original is better.
We can actually set minAvailable directly (thats the only use of maxUnavailable)

nikhiljindal replied:
oh actually, the bug seems to be in line 233.
Shouldnt it be:
`minAvailable := desired - maxUnavailable`

For ex: original = 30. desired = 2 and maxUnavailable = 50%.
Then minAvailable will be 29. It should be 1.

ironcladlou replied:
My memory's fuzzy but I think we compute relative to `original` as a compromise around asymmetry: with original=0, desired=10, maxUnavailable=50% we'd never scale down due to a minAvailable requirement of 5.

I added another commit which just sets maxUnavailable and minAvailable directly when desired=0 to remove all ambiguity. I'm inclined to not touch the minAvailable calculation otherwise at the moment since all the edge cases aren't in my head.

nikhiljindal replied:
That fixes the desired = 0 case.
original = 30. desired = 2 and maxUnavailable = 50% is still broken.

In your case, original=0, desired=10, maxUnavailable=50%, yes we will not scale down which is correct. There is no old pod to scale down :)

I hope that the tests will fail if we miss a case :)

ironcladlou replied:
I see what you mean now, I think `minAvailable := desired - maxUnavailable` is correct. The existing tests around that are incorrect, I think.

ironcladlou replied:
Fixed, new test cases and fixed some old ones for these asymmetrical scenarios.

nikhiljindal commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -208,9 +208,15 @@ func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
+	if desired == 0 {
+		maxUnavailable = int(math.Max(float64(oldRc.Spec.Replicas), float64(oldRc.Status.Replicas)))
+	} else {
+		if maxUnavailable == 0 && maxSurge == 0 {
```
fwiw, I was trying to find if the same problem exists with deployments and it does not because we do not check the computed values of maxSurge and maxUnavailable there, but check the original values instead
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/extensions/validation/validation.go#L263.

nikhiljindal commented:
I realize now that I was trying to reproduce with a deployment with replicas=0 where this issue does not exist. The issue is only with kubectl rolling-update.

nikhiljindal commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -208,13 +208,20 @@ func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
 	// The minumum pods which must remain available througout the update
 	// calculated for internal convenience.
-	minAvailable := original - maxUnavailable
+	minAvailable := desired - maxUnavailable
```
Sorry should have pointed it out earlier: We should set minAvailable to min(0, desired - maxUnavailable).

ironcladlou replied:
Wouldn't it be max(0, desired-maxUnavailable)?

ironcladlou replied:
(I assume the intent is to guard against negative values; if min is used, then minAvailable would always be 0 or negative)

nikhiljindal replied:
haha, ofcourse, sorry :)

nikhiljindal commented:
cc @janetkuo

nikhiljindal commented:
LGTM, thanks!

Please squash commits

ironcladlou commented:
Squashed, thanks

nikhiljindal commented:
Thanks, LGTM
